scopus/author_retrieval.py

```diff
@@ -179,10 +180,15 @@ def subject_areas(self):
         """
         areas = []
         area = namedtuple('Subjectarea', 'area abbreviation code')
-        for item in self._json['subject-areas'].get('subject-area', []):
-            new = area(area=item['$'], code=item['@code'],
-                       abbreviation=item['@abbrev'])
-            areas.append(new)
+        try:
+            for item in self._json.get('subject-areas',
+                                       {}).get('subject-area', []):
+                new = area(area=item['$'], code=item['@code'],
+                           abbreviation=item['@abbrev'])
+                areas.append(new)
+        except AttributeError:
+            # Sometimes subject-areas is mapped to None
+            pass
```

Comment:
The try-except clause is not necessary when you have `get('subject-areas', {})`. If "subject-areas" is not in `self._json`, it will return an empty dict and the next part will return an empty list.

Reply:
The problem is that for (at least) author 25823668100, the key 'subject-areas' does exist, but its value is None. Observe:

```
In [2]: AuthorRetrieval(25823668100, refresh=True).subject_areas

[...stack trace...]

In [4]: pdb.post_mortem(sys.last_traceback)
(Pdb) pp self._json['subject-areas']
None
(Pdb) pp type(self._json.get('subject-areas', {}))
<class 'NoneType'>
```
Perhaps a cleaner approach would be to make sure we never return keys for which we have no value in the first place, and then use the empty dict default here as you suggest.
I have to look at how the xml to json conversion is implemented.

Reply:
Ok, I looked into this a little further, and it's just that requests returns a dictionary that has key='subject-areas' and value=None.
```
url = 'https://api.elsevier.com/content/author/author_id/25823668100'
header = {'X-ELS-APIKey': 'xxxxxx', 'Accept': 'application/json'}
params = {'author_id': '25823668100', 'view': 'ENHANCED'}
res = requests.get(url, headers=header, params=params)
_json = json.loads(res.content)
response = _json['author-retrieval-response'].pop()

In [111]: type(response.get('subject-areas'))                                                                                                                                                 
Out[111]: NoneType

In [112]: type(response.get('subject-areas', {}))                                                                                                                                             
Out[112]: NoneType
```
I think that either we use this very ad-hoc try/catch to handle this case, or cleanup the returned json beforehand, as to avoid keys having an associated value that is None. Something like:
```
def customfilter(dct): 
    return {k:v for k,v in dct.items() if v is not None}

_json = json.loads(res.content, object_hook=customfilter)                                                                              
response = _json['author-retrieval-response'].pop()                                                                                                                                 

In [116]: type(response.get('subject-areas'))                                                                                                                                                 
Out[116]: NoneType

In [117]: type(response.get('subject-areas', {}))                                                                                                                                             
Out[117]: dict
```

Reply:
Thanks for the testing, that's most valuable.  It is indeed very strange that Scopus populates information that is None instead of just dropping the key.
I prefer the the try/except (as it was stated in the review which I just cancelled). First, it is a normal python workflow, to pretend something will work and handle exceptions accordingly. This is cheaper than checks. Second, rebuilding a dict is quite expensive and in most cases might not even be necessary. I would estimate that in 99% of cases the value we are looking for exist, so that the iterated dict from `customfilter()` will stay the same. And third, I am working on something else, namely a custom function parsing the dictionary tree without using all the get statements - it will simply work or not, and when it doesn't work, it will return None.

Reply:
Agreed, the exception will probably be triggered in rare cases, although the response time of the API dwarfs any dictionary reconstruction so in terms of performance it probably doesn't matter either way.
I will rebase everything on top of your latest commits and update this pull request.

Reply:
Thanks a lot @aieri!